### ISSUE TITLE
MapEntry objects cannot be equal to each other

### ISSUE THREAD

**Issue by srawlins** [dart-lang/sdk#32559]
MapEntry objects cannot be equal to each other (unless they're identical), which makes testing hard, for one thing.

```dart
var a = {'A': 1, 'B': 2};
var entries = a.entries;
expect(entries, contains(new MapEntry('A', 1))); // nope
expect(entries.singleWhere((e) => e.key == 'A').value, equals(1)); // yuck!
```

**Comment by lrhn** [comment]
I'ts probably useful. The class does scream "value type".

**Comment by davidmorgan** [comment]
@lrhn +1, this is surprising, and as Sam points out, makes testing hard.

**Comment by lrhn** [comment]
It's not a completely trivial feature, there is some design space. Too much to actually make this a slam-dunk feature.

Should a `MapEntry<int, int>(0, 0)` be equal to a `MapEntry<num, num>(0, 0)`? 
To `MapEntry<double, double>(0.0, 0.0)`? 
And should `MapEntry<num, num>(0, double.nan)`  not be equal to itself?

That is, do we just do equality only on the values, or does the type matter too?
That matters for testing, you can get false positives for equality if you don't consider these equal, but the class does. On the other hand, making the type matter may also give false negatives, when you compare your JSON map to a typed map literal in testing.

If the map uses a non-default equality and/or hashCode, should the `MapEntry` do the same for the key?
Most likely, otherwise you get into iffy behavior, say `map.entries.toSet().length != map.length`. 

However, if we do that, it prevents re-using an internal "entry" structure as the public entry because the internal structure would not need to remember the equality (the map already knows). It would also complicate (and enlarge) the `MapEntry` object to potentially include `hash` and `equals` and `isValidKey` functions (we'd need the last one in case we compare entries from different maps - which we can).

I guess this boils down to: Whether two objects are equal depends on your concept of equality. There is no *natural* equality that `MapEntry`, because there isn't one for `Map`, that we can safely assume. All choices can mismatch user expectation in some way. In cases like that, I usually prefer to keep equality *external* to the class.

We can add a `MapEntryEquality` to `package:collection`'s `Equality` classes, and even put `Equality<MapEntry<K, V>> get entryEquality;` on the `MapEquality` class. That should allow you to do any of the equality variants if you want to, without encumbering the `MapEntry` class with a single arbitrary and/or inefficient choice.

**Comment by lrhn** [comment]
Funny story ... it's *not* what Java does. Java has one non-default-equality map type, the `IdentityHashMap`. It's [entries](https://docs.oracle.com/javase/7/docs/api/java/util/IdentityHashMap.html#entrySet()) use identity instead of equality on *both* key and value.

A `MapEntry` is not just a pair. It's an entry of a map and should act as an entry of *that* map. I'd just prefer not to add operations that will make the entries more expensive. 

Asking for object equality on a composite object just for testing, is usually a waste of time. It's not going to be _sufficient_ in the long run. You want to be able to do more complicated tests on the sub-components than just equality.

I recommend either making a proper `Matcher` for map entries, rather than the rather simplistic point-wise comparison that this request would give you. 
```dart
Matcher entryWhere<K, V>({Object key, Object value}) => 
    _MapEntryMatcher<K, V>(wrapMatcher(key), wrapMatcher(value));
class _MapEntryMatcher<K, V> extends Matcher {
  final Matcher _key, _value;
  _MapEntryMatcher(this._key, this._value);

  bool matches(Object item, Map matchState) =>
    item is MapEntry<K, V> && _key.matches(item.key, matchState) && 
        _value.matches(item.value, matchState);

  Description describe(Description description) {
    description.add("MapEntry(");
    _key.describe(description);
    description.add(", ");
    _value.describe(description);
    description.add(")");
    return description;
  }
  // TODO: describe the mismatch somehow.
}
```
(or however `package:matcher` is currently designed) and then just write:
```dart
expect(map.entries, contains(entryWhere(key: 'A', value: 1)));
```
That's far more flexible than comparing structured data by only equality, and it's what the list/map matchers already do.

Or maybe get the test package to *recognize* `MapEntry` the same way that it recognizes `List` and `Map` in its recursive deep matcher, so you can direcly write:
```dart
expect(map.entries, contains(MapEntry('A', greaterThan(0)))); 
```
like you can currently write:
```
expect(list, [greaterThan(1), greaterThan(2)]);
```

**Comment by davidmorgan** [comment]
Thanks for correcting my Java trivia :)

Good point about matchers. Filed over at `package:matcher`.

**Comment by marcglasberg** [comment]
> Asking for object equality on a composite object just for testing, is usually a waste of time. 

Yes, but it's not just for testing. This complicates most operations with `MapEntry`. For example, I have a bunch of entries, and I want to put them in a Set, so that they become unique. I can't. 


> I guess this boils down to: Whether two objects are equal depends on your concept of equality. 

Yes, but most of the time you just need `key==other.key` and `value==other.value`. Surely, 95% of the time it's what you want.  This very obvious equality should be provided. 


> There is no natural equality that MapEntry, because there isn't one for Map, that we can safely assume. 

I see, for example, that built_collection does have comparable and hashable collections. So they assume that there is a natural equality. I guess the "natural equality" concept just means it's the most obvious one. I does not imply it's the only possible one, or that it is _sufficient_ in any way.

**Comment by morrica** [comment]
@irhn, call me Linus, but here's a rant. I agree with @marcglasberg. Testing is not the only use case for this or probably even the most common one. It's just the one you run into first when you are writing your tests. I'm implementing a `SortedMap` that operates using a `Comparator<MapEntry>` and a `SplayTreeSet<MapEntry>`. I need the set to be able to tell when it's dealing with functionally equivalent entries. I don't think the suggestions about `MapEntryEquality` or `Matcher` address that issue.

I find the design questions trivial to answer. They basically boil down to, "Are you going to include a runtimeType check in the value equality check?"
> Should a MapEntry<int, int>(0, 0) be equal to a MapEntry<num, num>(0, 0)?

With runtimeType check? Yes. Without? Also yes.
>To MapEntry<double, double>(0.0, 0.0)?

With runtimeType check? No. Without? Yes.
>And should MapEntry<num, num>(0, double.nan) not be equal to itself?

double.nan == double.nan **is** false, but have you ever seen a value equality check that didn't check identity first? I can't recall one off hand. Value checks are generally more expensive after all. So no, it would not, not be equal to itself.

>If the map uses a non-default equality and/or hashCode, should the MapEntry do the same for the key?
Most likely, otherwise you get into iffy behavior, say `map.entries.toSet().length != map.length`

Most likely not! If a developer is using non-default equality and/or hashCode **AND** relying on `MapEntry` equality by value then _that_ is the abnormal case and that developer should deal with the consequences in their design.

There _is_ a special case in Java with the `IdentityHashMap`, but the default is what everybody expects. Namely if `key1.equals(key2) && value1.equals(value2)` the entries are equal:
https://docs.oracle.com/en/java/javase/11/docs/api/java.base/java/util/Map.Entry.html#equals(java.lang.Object)

Is there a case now where developers are relying on the identity equality of `MapEntry`? Is it used in the SDK? If not, it seems like overriding == with a value check would be simple and helpful. And even if it is used in a few places it can be easily replaced with `identical(this, other)`.

Here's what Intellij thinks a good starting point should be:
```
  @override
  bool operator ==(Object other) =>
      identical(this, other) ||
      other is MapEntry &&
          runtimeType == other.runtimeType &&
          key == other.key &&
          value == other.value;
```
Might as well do the hashCode to be on the safe side too:
```
  @override
  int get hashCode => key.hashCode ^ value.hashCode;
```
This would make developers lives easier. Is there a reason that shouldn't be done?

**Comment by lrhn** [comment]
I'll never use `runtimeType` for equality, or anything else for that matter. That is probably a good thing since doing so disagrees with the rest of what you wrote: `MapEntry<int, int>(0, 0)` would not be equal to `MapEntry<num, num>(0, 0)` with a runtime check, and the hash code should probably include `runtimeType.hashCode` (not required, just fewer hash collisions).

I might drop the `identical` as well. It's an extra check which only pays off when it matches, otherwise it's extra overhead.
If you expect to match identity often, then you're likely depending on it and would be better off always using `identical` checks instead of `==`.
It's a *sound* short-curcuit check, but can be confusing when you have values like `NaN`. It means `MapEntry(0, nan)` is equal to itself, but not to another `MapEntry(0, nan)`, even with the same `nan` value (there can be more than one NaN value). The `identical` check is not *just* an optimization, it actually changes the the equality relation. Whether you want that or not is a design question for the particular use-case, not something we can definitely say for sure.
It does ensure reflexivity, though, but ensuring reflexivity is less impressive when NaN has already broken it.

So, I'd probably prefer something simple like:
```dart
bool operator==(Object other) => other is MapEntry && _key == other.key && value == other.value;
int get hashCode => _hash2(_key, value);
```

**Comment by ahmednfwela** [comment]
this is also causing issues in flutter https://github.com/flutter/flutter/issues/49670 if you make a dropdown from a map, using MapEntry as the value.

In general I would expect a `MapEntry` to behave exactly as a two-valued tuple.

```dart
MapEntry('a', 'b') == MapEntry('a', 'b') // should be true, similar to
('a','b') == ('a','b')
```

**Comment by lrhn** [comment]
Stove this issue was made, `MapEntry` has been made a `final` class, and `Map` implementations are no longer expected to use their internal representation directly as `MapEntry` objects. 

It has become closer to "just a pair", and more separate from the `Map` it comes from, and intentionally so. That makes it less problematic to implement `==`.
Might be worth reconsidering.
